packages/fast-usdc/src/exos/liquidity-pool.js

```diff
@@ -0,0 +1,185 @@
+  return zone.exoClassKit(
+    'Fast Liquidity Pool',
+    {
+      feeSink: M.interface('feeSink', {
+        receive: M.call(AmountShape, PaymentShape).returns(M.promise()),
+      }),
+      external: M.interface('external', {
+        setShareWorth: M.call(RatioShape).returns(M.promise()),
+      }),
+      depositHandler: M.interface('depositHandler', {
+        handle: M.call(SeatShape, M.any()).returns(M.promise()),
+      }),
+      withdrawHandler: M.interface('withdrawHandler', {
+        handle: M.call(SeatShape, M.any()).returns(M.promise()),
+      }),
+      public: M.interface('public', {
+        makeDepositInvitation: M.call().returns(M.promise()),
+        makeWithdrawInvitation: M.call().returns(M.promise()),
+        getPublicTopics: M.call().returns(TopicsRecordShape),
+      }),
+    },
+    /**
+     * @param {ZCFMint<'nat'>} shareMint
+     * @param {StorageNode} node
+     */
+    (shareMint, node) => {
+      const { brand: PoolShares } = shareMint.getIssuerRecord();
+      const { USDC } = brands;
+      const proposalShapes = makeProposalShapes({ USDC, PoolShares });
+      const dust = AmountMath.make(USDC, 1n);
+      const shareWorth = makeParity(dust, PoolShares);
+      const { zcfSeat: poolSeat } = zcf.makeEmptySeatKit();
+      const shareWorthRecorderKit = tools.makeRecorderKit(
+        node,
+        /** @type {TypedPattern<ShareWorth>} */ (RatioShape),
+      );
+      return {
+        shareMint,
+        shareWorth,
+        poolSeat,
+        PoolShares,
+        proposalShapes,
+        shareWorthRecorderKit,
+      };
+    },
+    {
+      feeSink: {
+        /**
+         * @param {Amount<'nat'>} amount
+         * @param {Payment<'nat'>} payment
+         */
+        async receive(amount, payment) {
+          const { poolSeat, shareWorth } = this.state;
+          const { external } = this.facets;
+          await depositToSeat(
+            zcf,
+            poolSeat,
+            harden({ USDC: amount }),
+            harden({ USDC: payment }),
+          );
+          await external.setShareWorth(withFees(shareWorth, amount));
+        },
+      },
+
+      external: {
+        /** @param {ShareWorth} shareWorth */
+        async setShareWorth(shareWorth) {
+          const { recorder } = this.state.shareWorthRecorderKit;
+          this.state.shareWorth = shareWorth;
+          await recorder.write(shareWorth);
+        },
+      },
+
+      depositHandler: {
+        /** @param {ZCFSeat} lp */
+        async handle(lp) {
+          const { shareWorth, shareMint, poolSeat } = this.state;
+          const { external } = this.facets;
+
+          /** @type {USDCProposalShapes['deposit']} */
+          // @ts-expect-error ensured by proposalShape
+          const proposal = lp.getProposal();
+          const post = depositCalc(shareWorth, proposal);
+          // COMMIT POINT
+          const mint = shareMint.mintGains(post.payouts);
```

Comment:
Right after minting the shareWorth has changed, even if the atomic rearrangement fails somehow.

Reply:
The shareWorth changes in a way that `depositCalc` anticipates. `post` says how much to mint.

`depositCalc` guarantees that the atomic rearrangement can't fail. If it fails, something very, very bad has happened. I suppose aborting the contract is the thing to do in that case. I can add a `try { ... } catch (oops=> zcf.shutdownWithFailure(oops) }`, following the [Prepare Commit](https://github.com/Agoric/agoric-sdk/wiki/Prepare-Commit) pattern. But the contract holds assets.

I suppose moving the assignment before the rearrange is easy.

Reply:
> depositCalc guarantees that the atomic rearrangement can't fail.

on closer review, I found that wasn't quite self-evident. I added some stuff to make it more clear.